### PR TITLE
NN-3895: Add incident role support to the draft adjudications

### PR DESCRIPTION
**Pull request by sp-watson**
I will add it to Reported Adjudications in a separate PR

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/DraftAdjudicationController.kt [hunk 1]
```diff
@@ -53,7 +56,9 @@ data class EditIncidentDetailsRequest(
   val locationId: Long? = null,
   @ApiModelProperty(value = "Date and time the incident occurred", example = "2010-10-12T10:00:00")
   @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
-  val dateTimeOfIncident: LocalDateTime? = null
+  val dateTimeOfIncident: LocalDateTime? = null,
+  @ApiModelProperty(value = "Information about the role of this prisoner in the incident")
+  val incidentRole: IncidentRoleDto? = null,
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/DraftAdjudicationController.kt:61]: Optional for now, but that will change once I update the UI code

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt [hunk 2]
```diff
@@ -30,6 +32,13 @@ data class IncidentDetailsDto(
   val handoverDeadline: LocalDateTime,
 )
 
+@ApiModel(value = "Incident Role")
+data class IncidentRoleDto(
+  @ApiModelProperty(value = "The incident role code", notes = "If not set then it is assumed they committed the offence on their own", example = "25a")
+  val roleCode: String?,
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt:38]: Will be null when the offender did it all by themselves

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt [hunk 3]
```diff
@@ -16,6 +16,8 @@ data class ReportedAdjudicationDto(
   val dateTimeReportExpires: LocalDateTime,
   @ApiModelProperty(value = "Incident details")
   val incidentDetails: IncidentDetailsDto,
+  // @ApiModelProperty(value = "Information about the role of this prisoner in the incident")
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt:19]: I will uncomment this in the next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/ReportedAdjudication.kt [hunk 4]
```diff
@@ -15,5 +15,7 @@ data class ReportedAdjudication(
   var locationId: Long,
   var dateTimeOfIncident: LocalDateTime,
   var handoverDeadline: LocalDateTime,
+  // var incidentRoleCode: String?,
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/entities/ReportedAdjudication.kt:18]: Will be uncommented in the next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt [hunk 5]
```diff
@@ -44,6 +47,13 @@ class DraftAdjudicationService(
         dateTimeOfIncident = dateTimeOfIncident,
         handoverDeadline = dateCalculationService.calculate48WorkingHoursFrom(dateTimeOfIncident)
       ),
+      // Temporary code for backwards compatibility
+      incidentRole = incidentRole?.let {
+        IncidentRole(
+          roleCode = incidentRole.roleCode,
+          associatedPrisonersNumber = incidentRole.associatedPrisonersNumber,
+        )
+      } ?: IncidentRole(roleCode = null, associatedPrisonersNumber = null),
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt:56]: Temporary code - to be removed in next PR

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt [hunk 6]
```diff
@@ -87,13 +94,17 @@ class DraftAdjudicationIntTest : IntegrationTestBase() {
       .jsonPath("$.draftAdjudication.incidentDetails.dateTimeOfIncident").isEqualTo(testAdjudication.dateTimeOfIncidentISOString)
       .jsonPath("$.draftAdjudication.incidentDetails.handoverDeadline").isEqualTo(testAdjudication.dateTimeOfIncident.plusDays(2).format(DateTimeFormatter.ISO_DATE_TIME))
       .jsonPath("$.draftAdjudication.incidentDetails.locationId").isEqualTo(testAdjudication.locationId)
+    // To be added when reported adjudications support it
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/DraftAdjudicationIntTest.kt:97]: Temporary code - to be removed in next PR

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/ReportedAdjudicationIntTest.kt [hunk 7]
```diff
@@ -172,6 +172,9 @@ class ReportedAdjudicationIntTest : IntegrationTestBase() {
       .jsonPath("$.draftAdjudication.incidentDetails.dateTimeOfIncident").isEqualTo(IntTestData.DEFAULT_ADJUDICATION.dateTimeOfIncidentISOString)
       .jsonPath("$.draftAdjudication.incidentDetails.locationId").isEqualTo(IntTestData.DEFAULT_ADJUDICATION.locationId)
       .jsonPath("$.draftAdjudication.incidentDetails.handoverDeadline").isEqualTo(IntTestData.DEFAULT_HANDOVER_DEADLINE_ISO_STRING)
+      // To be added
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/integration/ReportedAdjudicationIntTest.kt:175]: Temporary code - to be sorted in next PR

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/ReportedAdjudicationServiceTest.kt [hunk 8]
```diff
@@ -195,6 +196,11 @@ class ReportedAdjudicationServiceTest {
         dateTimeOfIncident = DATE_TIME_OF_INCIDENT,
         handoverDeadline = DATE_TIME_REPORTED_ADJUDICATION_EXPIRES
       ),
+      // Temporary code until we implement in reported adjudications
```
Comment (sp-watson) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/ReportedAdjudicationServiceTest.kt:199]: Temporary code - to be sorted in next PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt [hunk 9]
```diff
@@ -14,6 +14,8 @@ data class DraftAdjudicationDto(
   val prisonerNumber: String,
   @ApiModelProperty(value = "Incident details")
   val incidentDetails: IncidentDetailsDto,
+  @ApiModelProperty(value = "Information about the role of this prisoner in the incident")
+  val incidentRole: IncidentRoleDto,
```
Comment (richardpopple) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt:18]: Does this need to be optional at the moment to?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt (hunk 2 above)
Comment (richardpopple) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt:38]: So does this mean that the IncidentRoleDto is effectively empty if the prisoner committed it on their own?

Comment (richardpopple) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/DraftAdjudicationDto.kt:38]: Am I also right in thinking that its one adjudication for each incident role - I think that's correct

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt (hunk 5 above)
Comment (richardpopple) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/DraftAdjudicationService.kt:56]: Will we need to blat the data that is already in the database - i.e. any draft adjudications without an IncidentRole?

**Review by richardpopple — APPROVED** [review]

**Pull request merged**